### PR TITLE
Remove jQuery usage from range.js

### PR DESCRIPTION
**Pull request by robertknight**
As part of the removal of jQuery from the annotator, remove usage from
`range.js`:

 - Replace `parents()` method with a small helper function

 - Rewrite `getTextNodes` function to accept and return DOM nodes rather
   than jQuery collections

 - Use `Node.contains` instead of jQuery's `contains` method. Note that there
   is a semantic difference that `contains(node, node)` returns
   `false` whereas `node.contains(node)` returns `true`. In other words, in the DOM API a node "contains" itself, but not in jQuery. It was possible to leverage this to simplify a condition.

### CODE REVIEW

File: src/annotator/anchoring/range.js [hunk 1]
```diff
@@ -423,17 +440,15 @@ export class SerializedRange {
...
-      .each(function () {
-        if (contains(this, range.endContainer)) {
-          range.commonAncestorContainer = this;
-          // bail out of loop
-          return false;
-        }
-        return true;
-      });
+    const contains = (a, b) =>
+      a.compareDocumentPosition(b) & Node.DOCUMENT_POSITION_CONTAINED_BY;
```
Comment (robertknight) [inline src/annotator/anchoring/range.js:444]: See https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition for enum names and associated values.

File: src/annotator/anchoring/range.js [hunk 2]
```diff
@@ -382,7 +396,10 @@ export class SerializedRange {
         targetOffset--;
       }
 
-      for (let tn of getTextNodes($(node))) {
+      for (let tn of getTextNodes(node)) {
+        if (tn.nodeValue === null) {
+          continue;
```
Comment (LMS007) [inline src/annotator/anchoring/range.js:401]: When is a text node's nodeValue null? Is it possible to add a test for this case?

File: src/annotator/anchoring/range.js [hunk 3]
```diff
@@ -423,17 +440,15 @@ export class SerializedRange {
...
-      .parents()
-      .each(function () {
-        if (contains(this, range.endContainer)) {
-          range.commonAncestorContainer = this;
-          // bail out of loop
-          return false;
-        }
-        return true;
-      });
+    const contains = (a, b) =>
```
Comment (LMS007) [inline src/annotator/anchoring/range.js:443]: The comment above still likely applies, but does reference jquery so we should probably at least change that or delete the parts of it that don't make sense anymore.

**Review by LMS007 — APPROVED** [review]
If you could just make a change to that comment below and decide on that missing unit test, then we are good. Its nice to get jquery removed!

File: src/annotator/anchoring/range.js (hunk 2 above)
Comment (robertknight) [reply src/annotator/anchoring/range.js:401]: I looked this up in the DOM specs this morning and found that `nodeValue` is always a string for a `Text` node: https://dom.spec.whatwg.org/#dom-node-nodevalue.

In the case of `Text` nodes `nodeValue` is an alias for `data` which is always non-null. I've changed this to use `data` instead.

File: src/annotator/anchoring/range.js (hunk 3 above)
Comment (robertknight) [reply src/annotator/anchoring/range.js]: Agreed. I re-tested the test case given in the comment above and was no longer able to reproduce. I couldn't find any relevant bug reports in WebKit's bug tracker either. On that basis I simplified this code with `Node.contains(...)`.

In future I suspect we might be able to remove `Range` normalization from most of the code. The only place where it is really needed AFAIK is before wrapping nodes in highlights.

**Pull request merged**